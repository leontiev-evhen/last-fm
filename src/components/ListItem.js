import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Image } from './Image';

export function ListItem({ track }) {
	return (
		<div className="col">
			<div className="track-section">
				<div className="title">
					<Link to={`/artist/${track.artist.name}`}>{track.artist.name}</Link>
					<div className="track-name">{track.name}</div>
				</div>
				<Image path={track.image.length > 3 ? track.image[3]['#text'] : undefined} />
			</div>
		</div>
	);
}

ListItem.propTypes = {
	track: PropTypes.shape({
		name: PropTypes.string.isRequired,
		image: PropTypes.array,
		artist: PropTypes.shape({
      name: PropTypes.string.isRequired,
    }).isRequired,
	}).isRequired,
};
