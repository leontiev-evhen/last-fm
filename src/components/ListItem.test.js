import React from 'react';
import { shallow } from 'enzyme';
import { ListItem } from './ListItem';
import { Link } from 'react-router-dom';

describe('ListItem component', () => {
	const props = {
		track: {
			name: 'track',
			image: [
				{
					'#text': 'image1',
				},
				{
					'#text': 'image2',
				},
				{
					'#text': 'image3',
				},
				{
					'#text': 'image3',
				},
			],
			artist: {
				name: 'artist',
			},
		},
	};
	const component = shallow(<ListItem {...props} />);

	it('render correctly ListItem component', () => {
		expect(component).toMatchSnapshot();
	});
	it('correct track name', () => {
		expect(component.find('.track-name').text()).toEqual(props.track.name);
	});
	it('correct Link to', () => {
		expect(component.find('Link').prop('to')).toEqual(
			`/artist/${props.track.artist.name}`
		);
	});
	it('correct Link', () => {
		expect(component.find('Link')).toHaveLength(1);
	});
	it('Link contains correct value', () => {
		expect(
			component.contains(
				<Link to={`/artist/${props.track.artist.name}`}>
					{props.track.artist.name}
				</Link>
			)
		).toBeTruthy();
	});
	it('correct Image src', () => {
		expect(component.find('Image').prop('path')).toEqual(
			props.track.image[3]['#text']
		);
	});
	describe('Image array size is less than three', () => {
		const nextProps = {
			track: {
				...props.track,
				image: [
					{
						'#text': 'image1',
					},
					{
						'#text': 'image2',
					},
					{
						'#text': 'image3',
					},
				],
			},
		};
		const component = shallow(<ListItem {...nextProps} />);
		it('Image component path is undefined', () => {
			expect(component.find('Image').prop('path')).toEqual(undefined);
		});
	});
});
