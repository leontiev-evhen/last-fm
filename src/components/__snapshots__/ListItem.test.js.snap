// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component render correctly ListItem component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListItem
    track={
      Object {
        "artist": Object {
          "name": "artist",
        },
        "image": Array [
          Object {
            "#text": "image1",
          },
          Object {
            "#text": "image2",
          },
          Object {
            "#text": "image3",
          },
          Object {
            "#text": "image3",
          },
        ],
        "name": "track",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="track-section"
      >
        <div
          className="title"
        >
          <Link
            replace={false}
            to="/artist/artist"
          >
            artist
          </Link>
          <div
            className="track-name"
          >
            track
          </div>
        </div>
        <Image
          path="image3"
        />
      </div>,
      "className": "col",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="title"
          >
            <Link
              replace={false}
              to="/artist/artist"
            >
              artist
            </Link>
            <div
              className="track-name"
            >
              track
            </div>
          </div>,
          <Image
            path="image3"
          />,
        ],
        "className": "track-section",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                replace={false}
                to="/artist/artist"
              >
                artist
              </Link>,
              <div
                className="track-name"
              >
                track
              </div>,
            ],
            "className": "title",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "artist",
                "replace": false,
                "to": "/artist/artist",
              },
              "ref": null,
              "rendered": "artist",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "track",
                "className": "track-name",
              },
              "ref": null,
              "rendered": "track",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "path": "image3",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="track-section"
        >
          <div
            className="title"
          >
            <Link
              replace={false}
              to="/artist/artist"
            >
              artist
            </Link>
            <div
              className="track-name"
            >
              track
            </div>
          </div>
          <Image
            path="image3"
          />
        </div>,
        "className": "col",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="title"
            >
              <Link
                replace={false}
                to="/artist/artist"
              >
                artist
              </Link>
              <div
                className="track-name"
              >
                track
              </div>
            </div>,
            <Image
              path="image3"
            />,
          ],
          "className": "track-section",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Link
                  replace={false}
                  to="/artist/artist"
                >
                  artist
                </Link>,
                <div
                  className="track-name"
                >
                  track
                </div>,
              ],
              "className": "title",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "artist",
                  "replace": false,
                  "to": "/artist/artist",
                },
                "ref": null,
                "rendered": "artist",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "track",
                  "className": "track-name",
                },
                "ref": null,
                "rendered": "track",
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "path": "image3",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
